### PR TITLE
M3 fix enabling queue

### PR DESCRIPTION
**Pull request by hluchas**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/8240
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
When you switch between RabbitMQ and Beanstalk in queue config, you'll get exception.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Try to change queue settings in config and select between RabbitMQ and Beanstalk
2. You'll get this
```
Symfony\Component\DependencyInjection\Exception\LogicException: You cannot register extension "OldSound\RabbitMqBundle\DependencyInjection\OldSoundRabbitMqExtension" from "Mautic\QueueBundle\DependencyInjection\MauticQueueExtension". Extensions must be registered before the container is compiled.
```

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Use reproduction steps

### CODE REVIEW

File: app/bundles/QueueBundle/MauticQueueBundle.php [hunk 1]
```diff
@@ -11,11 +11,82 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function build(ContainerBuilder $container)
+    {
+        parent::build($container);
+
+        $queueProtocol = $this->getQueueProtocol();
+
+        if (QueueProtocol::RABBITMQ === $this->getQueueProtocol()) {
```
Comment (dongilbert) [inline app/bundles/QueueBundle/MauticQueueBundle.php]: Not a big deal, but can we use $queueProtocol here?

File: app/bundles/QueueBundle/MauticQueueBundle.php [hunk 2]
```diff
@@ -11,11 +11,82 @@
...
+        }
+
+        if (QueueProtocol::BEANSTALKD === $queueProtocol) {
+            return new LeezyPheanstalkExtension();
+        }
+    }
+
+    private function getQueueProtocol(): string
+    {
+        if (empty($this->localParams['queue_protocol'])) {
```
Comment (dongilbert) [inline app/bundles/QueueBundle/MauticQueueBundle.php]: If we change this to `return $this->localParams['queue_protocol'] ?? null;` we can simplify the `getContainerExtension` method to `return $this->createContainerExtension()`

File: app/bundles/QueueBundle/MauticQueueBundle.php [hunk 3]
```diff
@@ -11,11 +11,82 @@
...
+            return null;
+        }
+
+        if (QueueProtocol::RABBITMQ === $queueProtocol) {
+            return new OldSoundRabbitMqExtension();
+        }
+
+        if (QueueProtocol::BEANSTALKD === $queueProtocol) {
+            return new LeezyPheanstalkExtension();
+        }
```
Comment (dongilbert) [inline app/bundles/QueueBundle/MauticQueueBundle.php]: Maybe ditch lines 71-73 and add a `return null;` here at the end.

**Comment by dongilbert** [comment]
I think the only other thing we need is a way to delete the container if the queue protocol changes so it can rebuild.

**Comment by alanhartless** [comment]
@dongilbert we'll have to remove the queue settings from the UI as "advanced config" changes which I'll do in my "containerless config" PR. People will have to be instructed to update their local.php directly and then delete the cache because Mautic will no longer do that.

**Comment by dongilbert** [comment]
What about creating a command that can set the value, for a better user experience. `bin/console queue:enable` and that will ask which one to use and then any settings needed, the update and clear the cache

File: app/bundles/QueueBundle/MauticQueueBundle.php (hunk 3 above)
Comment (hluchas) [reply app/bundles/QueueBundle/MauticQueueBundle.php]: I prefer returning strictly same type.

File: app/bundles/QueueBundle/MauticQueueBundle.php (hunk 1 above)
Comment (hluchas) [reply app/bundles/QueueBundle/MauticQueueBundle.php]: :-D

File: app/bundles/QueueBundle/MauticQueueBundle.php (hunk 3 above)
Comment (dongilbert) [reply app/bundles/QueueBundle/MauticQueueBundle.php]: If the queue protocol is set to something else besides beanstalk or rabbitmq then it won't return anything at all

File: app/bundles/QueueBundle/MauticQueueBundle.php (hunk 2 above)
Comment (hluchas) [reply app/bundles/QueueBundle/MauticQueueBundle.php]: I don't understand what do you mean. Could you provide piece of code pls?

Comment (dongilbert) [reply app/bundles/QueueBundle/MauticQueueBundle.php]: It's ok - no need to.

**Comment by hluchas** [comment]
Okay, this is still producing the same error.  Moving to pending feedback.

**Pull request merged**